src/ISDeviceCal.cpp

```diff
@@ -0,0 +1,822 @@
+        }
+
+        // Accel Data
+        key = formatString("acc%d", d+1);
+        if (jCal.contains(key))
+        {
+			if (ocal) OrthoLeastSquaresDataJSONToSensor(jCal[key], ocal->acc[d]);
+            if (mcal && mcal->acc[d].orth) OrthoJSONToSensor(jCal[key], mcal->acc[d].orth, mcal->acc[d].bias);
+        }
+    }
+
+    for (int d = 0; d < NUM_MAG_DEVICES; d++)
+    {   // Mag Data
+        // Set default identity matrix
+        for (int i = 0; i < 9; i++)
+        {
+            float val = ((i == 0 || i == 4 || i == 8) ? 1.0f : 0.0f);
+            mcal->mag[d].orth[i] = val;
+        }
+
+        std::string key = formatString("mag%d", d+1);
+        if (jCal.contains(key))
+        {
+			if (ocal) OrthoLeastSquaresDataJSONToSensor(jCal[key], ocal->mag[d]);
+            if (mcal && mcal->mag[d].orth) OrthoJSONToSensor(jCal[key], mcal->mag[d].orth, mcal->mag[d].bias);
+        }
+    }
+}
+
+json ISDeviceCal::saveMcToJSON(const std::string& filePath, int pose, sOrthoCal *cal, sensor_mcal_group_t *mcal)
+{
+    json jCal = json::object();
+    jCal["comment_line1"] = formatString(" Sensor Calibration Data - %s ", filePath.c_str());
+    jCal["comment_line2"] = " Y = A * X.  Y = truth data, X = adc measured data, A = calibration data. ";
+
+    // Add last pose saved to xml.  If we crash, we can restart from here.
+    if (pose != -1)
+    {
+        jCal["curPose"] = pose;
+    }
+
+    for (int d = 0; d < NUM_IMU_DEVICES; d++)
+    {
+        // Gyro Data
+        if (motionCalPresent(mcal->pqr[d].orth, mcal->pqr[d].bias))
+        {
+            json jGyr = json::object();
+            if (cal)   jGyr = OrthoLeastSquaresDataSensorToJSON(jGyr, cal->gyr[d]);
+            if (mcal && mcal->pqr[d].orth)  jGyr = OrthoSensorToJSON(jGyr, mcal->pqr[d].orth, mcal->pqr[d].bias);
+            jCal[formatString("gyr%d", d + 1)] = jGyr;
+        }
+
+        // Accel Data
+        if (motionCalPresent(mcal->acc[d].orth, mcal->acc[d].bias))
+        {
+            json jAcc = json::object();
+            if (cal)   jAcc = OrthoLeastSquaresDataSensorToJSON(jAcc, cal->acc[d]);
+            if (mcal && mcal->acc[d].orth)  jAcc = OrthoSensorToJSON(jAcc, mcal->acc[d].orth, mcal->acc[d].bias);
+            jCal[formatString("acc%d", d + 1)] = jAcc;
+        }
+    }
+
+    for (int d = 0; d < NUM_MAG_DEVICES; d++)
+    {
+        // Mag Data
+        if (motionCalPresent(mcal->mag[d].orth, mcal->mag[d].bias))
+        {
+            json jMag = json::object();
+            if (cal)   jMag = OrthoLeastSquaresDataSensorToJSON(jMag, cal->mag[d]);
+            if (mcal && mcal->mag[d].orth)  jMag = OrthoSensorToJSON(jMag, mcal->mag[d].orth, mcal->mag[d].bias);
+            jCal[formatString("mag%d", d + 1)] = jMag;
+        }
+    }
+
+    return jCal;
+}
+
+int ISDeviceCal::uploadSensorCalStep(port_handle_t port, int &calUploadState, sensor_cal_t &cal)
```

Comment:
This really seems more appropriate for ISDevice::step() if a calibration upload is being performed, but its probably "sixes".
Also, this shouldn't be using `port_handle_t`, and `comManagerSendData()` but instead using `ISDevice->sendData()`